### PR TITLE
Wrap blockmode

### PR DESCRIPTION
**Pull request by arleaman**
**Description of proposed changes**

Wrap blockmode

Addresses #1091

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [x] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/src/blockm.py [hunk 1]
```diff
@@ -14,18 +14,18 @@
...
-    Block average (x,y,z) data tables by mean or median estimation.
+    Block average (x,y,z) data tables by mean, median, or mode estimation.
 
     Reads arbitrarily located (x,y,z) triples [or optionally weighted
-    quadruples (x,y,z,w)] from a table and writes to the output a mean or
-    median (depending on ``block_method``) position and value for every
+    quadruples (x,y,z,w)] from a table and writes to the output a mean,
+    median, or mode (depending on ``block_method``) position and value for every
     non-empty block in a grid region defined by the ``region`` and ``spacing``
     parameters.
```
Comment (meghanrjones) [inline pygmt/src/blockm.py:23]:
```suggestion
    median, or mode (depending on ``block_method``) position and value for
    every non-empty block in a grid region defined by the ``region`` and
    ``spacing`` parameters.
```
Just a small formatting fix for line-length.

File: pygmt/src/blockm.py [hunk 2]
```diff
@@ -14,18 +14,18 @@
...
 
     Parameters
     ----------
     block_method : str
-        Name of the GMT module to call. Must be "blockmean" or "blockmedian".
+        Name of the GMT module to call. Must be "blockmean" "blockmedian" or "blockmode".
```
Comment (meghanrjones) [inline pygmt/src/blockm.py:28]:
```suggestion
        Name of the GMT module to call. Must be "blockmean" "blockmedian" or
        "blockmode".
```
Another small formatting fix.

**Review by meghanrjones — COMMENTED** [review]
Great work, @arleaman! I made a couple small formatting suggestions. Apart from that, there are two more steps needed before adding this module to PyGMT:

1. Add `blockmode` to `/doc/api/index.rst` in line 84.
2. Add a test for blockmode. I can provide some instructions for how to add a test for this module, if you expect to have time to work with us on finishing up the pull request?

**Comment by seisman** [comment]
Ping @arleaman to finalize this PR if available.

File: pygmt/src/blockm.py [hunk 3]
```diff
@@ -14,18 +14,19 @@
...
+    quadruples (x,y,z,w)] from a table and writes to the output a mean,
+    median, or mode (depending on ``block_method``) position and value for
+    every non-empty block in a grid region defined by the ``region`` and
+    ``spacing`` parameters.
 
     Parameters
     ----------
     block_method : str
-        Name of the GMT module to call. Must be "blockmean" or "blockmedian".
+        Name of the GMT module to call. Must be "blockmean" "blockmedian" or
```
Comment (seisman) [inline pygmt/src/blockm.py:28]:
```suggestion
        Name of the GMT module to call. Must be "blockmean", "blockmedian" or
```

File: pygmt/src/blockm.py [hunk 4]
```diff
@@ -210,3 +211,74 @@ def blockmedian(table=None, outfile=None, *, x=None, y=None, z=None, **kwargs):
...
+    V="verbose",
+    a="aspatial",
+    f="coltypes",
+    i="incols",
+    o="outcols",
+    r="registration",
+    s="skiprows",
+    w="wrap",
+)
+@kwargs_to_strings(R="sequence")
```
Comment (seisman) [inline pygmt/src/blockm.py:229]:
```suggestion
@kwargs_to_strings(R="sequence", i="sequence_comma")
```

**Review by seisman — APPROVED** [review]
Looks good to me except two minor suggestions.

**Review by weiji14 — APPROVED** [review]
:tada:

**Comment by arleaman** [comment]
Thanks @meghanrjones! I'll be happy to add tests for the module with some additional guidance!
> 2. Add a test for blockmode. I can provide some instructions for how to add a test for this module, if you expect to have time to work with us on finishing up the pull request?

**Comment by meghanrjones** [comment]
Thanks, @arleaman! Before adding tests, let's check with the other reviewers whether a test is needed for this new function or the tests for blockmedian and blockmode provide sufficient coverage. What do you think, @weiji14 and @seisman?

**Comment by seisman** [comment]
> Thanks, @arleaman! Before adding tests, let's check with the other reviewers whether a test is needed for this new function or the tests for blockmedian and blockmean provide sufficient coverage. What do you think, @weiji14 and @seisman?

I think we still need one test for blockmode to make sure that the function can be called.

**Comment by meghanrjones** [comment]
Great, let's add a test! @arleaman, here is some guidance:

1. Since the modules blockmean and blockmode are so similar, we can add the test to the file with the blockmean tests rather than creating a new file with separate imports/datasets for blockmode. To do this, we need to first generalize the blockmean tests filename (currently `pygmt/tests/test_blockmean.py`):
   ```
   cd <path-to-pygmt-repository>
   git checkout wrap-blockmode
   git mv pygmt/tests/test_blockmean.py pygmt/tests/test_blockm.py # Renames the file
   git status # Check that the change is ready to be committed
   git commit -m "Rename test_blockmean.py to test_blockm.py"
   ```
2. Generalize the docstring in `pygmt/tests/test_blockm.py` by updating L2 "`Tests for blockmean.`" to "`Tests for blockmean and blockmode.`"
3. Add `blockmode` to the `from pygmt import blockmean` statement on L9
4. Create a new test for blockmode. You can use the same structure one of the tests for blockmean:
   - Copy/paste the first test in `pygmt/tests/test_blockm.py` (`test_blockmean_input_dataframe`) and change the name for the new function to `test_blockmode_input_dataframe`
   - Change the function call from `blockmean(table=dataframe...)` to `blockmode(table=dataframe...)`
   - Update the last line in the test (`npt.assert_allclose(...)`) so that the output is correct for `blockmode` rather than `blockmean` (i.e., change -384.0 to -385.0)

**Comment by arleaman** [comment]
> Great, let's add a test! @arleaman, here is some guidance:

Thanks for the help @meghanrjones!

File: pygmt/src/blockm.py [hunk 5]
```diff
@@ -210,3 +211,74 @@ def blockmedian(table=None, outfile=None, *, x=None, y=None, z=None, **kwargs):
...
     )
+
+
+@fmt_docstring
+@use_alias(
+    I="spacing",
+    R="region",
+    V="verbose",
+    a="aspatial",
+    f="coltypes",
```
Comment (meghanrjones) [inline pygmt/src/blockm.py:222]:
```suggestion
    b="binary",
    d="data",
    e="find",
    f="coltypes",
    h="header",
```

File: pygmt/src/blockm.py [hunk 6]
```diff
@@ -210,3 +211,74 @@ def blockmedian(table=None, outfile=None, *, x=None, y=None, z=None, **kwargs):
...
+    {I}
+
+    {R}
+
+    outfile : str
+        The file name for the output ASCII file.
+
+    {V}
+    {a}
+    {f}
```
Comment (meghanrjones) [inline pygmt/src/blockm.py:265]:
```suggestion
    {b}
    {d}
    {e}
    {f}
    {h}
```

**Review by meghanrjones — COMMENTED** [review]
No problem and thanks for making the changes! :smile:

I made a couple suggestions related to adding some new common options.

Also, could you please merge the latest changes into your branch by clicking `update branch`? 

image

File: pygmt/tests/test_blockm.py [hunk 7]
```diff
@@ -96,3 +96,14 @@ def test_blockmean_without_outfile_setting():
...
+
+def test_blockmode_input_dataframe(dataframe):
+    """
+    Run blockmode by passing in a pandas.DataFrame as input.
+    """
+    output = blockmode(table=dataframe, spacing="5m", region=[245, 255, 20, 30])
+    assert isinstance(output, pd.DataFrame)
+    assert all(dataframe.columns == output.columns)
+    assert output.shape == (5849, 3)
+    npt.assert_allclose(output.iloc[0], [245.888877, 29.978707, -385.0])
```
Comment (meghanrjones) [inline pygmt/tests/test_blockm.py:109]:
```suggestion
    npt.assert_allclose(output.iloc[0], [245.88819, 29.97895, -385.0])
```

I forget that the x, y results would also be different for blockmode relative to blockmean.

**Review by meghanrjones — APPROVED** [review]
Great work! :tada:

I'll leave this open for a short while in case others have any last minute comments.

**Review by seisman — APPROVED** [review]

**Pull request merged**